### PR TITLE
M33: Get long filename

### PR DESCRIPTION
**Pull request by thinkyhead**
- Proof of concept. Retrieve the long filename of a file or folder
  based on the DOS path
- Includes some nice improvements to GCode parsing

### CODE REVIEW

File: Marlin/configurator/config/language.h [hunk 1]
```diff
@@ -167,7 +167,8 @@
 #define MSG_ENDSTOP_OPEN                    "open"
 #define MSG_HOTEND_OFFSET                   "Hotend offsets:"
 
-#define MSG_SD_CANT_OPEN_SUBDIR             "Cannot open subdir"
+#define MSG_SD_CANT_OPEN_SUBDIR             "Cannot open subdir: "
```
Comment (thinkyhead) [inline Marlin/configurator/config/language.h:170]: @daid @repetier @foosel I propose this change in the serial output.

Comment (daid) [inline Marlin/configurator/config/language.h:170]: @thinkyhead no problem for me (not using any of the SD card functions from the host). But might cause problems for octoprint, as I think @foosel is using the SD card functions.

**Comment by thinkyhead** [comment]
@repetier @kliment @daid @foosel 
This feature will make your hosts' SD file listings look nicer, if you decide to include file listings or file browsers in your hosts. Also considering an extended `M20` as a faster method to get full listings.

File: Marlin/configurator/config/language.h (hunk 1 above)
Comment (foosel) [inline Marlin/configurator/config/language.h:170]: I am, but I'm not evaluating that message at all so there's not a problem.

**Comment by foosel** [comment]
The problem from my point of view with this PR is that it again _does way more than the packaging says_. The first commit alone changes five files in nearly 200 lines and most of those changes have nothing to do with the new M33 but instead refactor parts of the very sensitive command processing, and -- if my C is not so rusty I completely misunderstand the code -- actually introduce breaking changes. 

Refactoring should be done, yes, but hiding it like that in a new feature is counter intuitive and makes making sure that everything has been tested properly overly complicated.

Since this PR does WAY more than it advertises, my two cents would be to separate the M33 stuff from the rest -- and do some very rigorous testing to make 100% sure the rest doesn't break things.

**Comment by thinkyhead** [comment]
@foosel I'm getting pretty good at re-rolling PRs through cherry-picking. I don't mind doing that here, if it helps. Usually I put in several PRs a day that no one ever sees, so I don't usually get called on this. My point with this one is to get it out and in front of you for review, not to make it perfect yet. That will come.

**Comment by thinkyhead** [comment]
#2098 (pre-sanitize g-code) makes much of this obsolete – specifically the need to watch out for `N` and `*` in commands (and no `clear_asterisk` function needed). So that will be committed first, then this will be re-done.

**Comment by thinkyhead** [comment]
#2108 is the distilled feature. Other bits already split up into smaller PRs. My favorite new trick to make a starting-point for cherry-picking…!

```
git checkout Development -b shiny_new_branch
git diff Development branch_with_changes | git apply
```

**Pull request closed without merging**